### ISSUE TITLE
Ivy configuration mappings cause an incomplete set of variants to be reported

### ISSUE THREAD

**Issue by ghale** [gradle/gradle#24016]
When Ivy configuration mappings are present, they can cause the wrong variants to be selected when `ResolvedComponentResult.getDependenciesForVariant()` is called.

For example, given a set of dependencies with configuration mappings:

airlift:
depends on airlift-log-manager with the mapping compile -> compile,master
depends on airlift-log-manager with the mapping runtime -> compile,master,runtime

airlift-log-manager:
depends on log4j-over-slf4j with the mapping runtime -> compile, master, runtime

log4j-over-slf4j:
contains an artifact in the "master" conf but not compile or runtime.

So the only dependency to slf4j is through the runtime configuration in airlift-log-manager.

### Expected Behavior
Resolving the default dependency on airlift should show the all three variants, with the runtime dependency also containing a transitive dependency down to log4j-over-slf4j.

### Current Behavior
A default dependency on airlift should resolve to a set of artifacts that contains the log4j-over-slf4j, which it does.  However, when you look at the variants, which is what the build scan plugin does, we get an incomplete view of the variants.  The problem is that the Ivy mappings get translated into selectors on the airlift component, but we only allow one selector per dependency and the compile -> compile,master becomes the selector for the airlift-log-manager dependency.  This then means that the runtime variant of airlift-log-manager is not even considered because it doesn't match the selector.

When the build scan plugin queries the dependencies of airlift, it asks for the variants of airlift-log-manager and gets `compile` and `master`, neither of which have a path to log4j-over-slf4j so this doesn't show up in the build scan.

The second issue is that in `EdgeState.getSelectedVariant()` we always select the first matching variant, which means that even if the `compile` dependency is removed and the only dependency in airlift is `runtime->compile,master,runtime`, the `master` variant gets selected (`compile` is ignored because `runtime` extends `compile`).  This is logically wrong from the perspective of what Ivy mappings should do.  According to the [Ivy docs](https://ant.apache.org/ivy/history/latest-milestone/ivyfile/dependency.html) something like `runtime->compile,master,runtime` means the dependency is in the `runtime` configuration and should contain _all_ of the dependencies from the `compile`, `master` and `runtime` configurations of the dependency component, not just the first one that matches.  So I think there is some further modeling necessary to more accurately capture Ivy mappings as variants.

### Context
This showed up in a build scan where it only showed the compile and master variants and did not show runtime at all.

<img width="683" alt="image" src="https://user-images.githubusercontent.com/642656/220933540-b2064885-6e47-4b1f-b6c1-6b3e18927296.png">

The dependencies report (which does not show variants) accurately resolved the graph to contain all of the dependencies.  It's just when you query the variants that you get an incomplete view.

### Steps to Reproduce 
The following repo captures the dependency relationships:

[ivy-repo.zip](https://github.com/gradle/gradle/files/10814446/ivy-repo.zip)

Then, with a build script like this (replace the repo path with wherever the ivy repo has been placed):

```
plugins {
    id 'java-library'
}

repositories {
    ivy {
        url "file://${projectDir}/repo"
        patternLayout {
            ivy "[organisation]/[module]/[revision]/[module]-[revision].ivy"
            artifact "[organisation]/[module]/[revision]/[artifact]-[revision](-[classifier])(.[ext])"
            m2compatible = true
        }
    }
}

dependencies {
    api 'io.airlift:bootstrap:222'
}
```

You can run the `dependencies` task with `--scan` and you'll see the correctly resolved dependencies but the build scan will have the incomplete set of variants.

### Your Environment
Build scan URL: https://scans.gradle.com/s/orjo7nqpg7lpm/dependencies?toggled=W1swXSxbMCwxXSxbMCwxLFswXV0sWzAsMl0sWzAsMixbMF1dXQ

**Comment by big-guy** [comment]
@ljacomet were you going to look at this more?

**Comment by ljacomet** [comment]
@big-guy No, but as discussed we should still look into having support for such cases.
